cmd/list.go

```diff
@@ -1,42 +1,178 @@
 package cmd
 
 import (
+	"context"
 	"fmt"
 	"os"
 	"sort"
+	"strings"
 	"text/tabwriter"
 
-	"github.com/whalebrew/whalebrew/packages"
+	"github.com/Masterminds/semver"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/go-units"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"github.com/whalebrew/whalebrew/client"
+	"github.com/whalebrew/whalebrew/packages"
 )
 
 func init() {
 	RootCmd.AddCommand(listCommand)
 }
 
+type listInfo struct {
+	version      *semver.Version
+	imageSummary *types.ImageSummary
+	tag          string
+}
+
+type listInfos []*listInfo
+
+func (c listInfos) Len() int {
+	return len(c)
+}
+
+// Less is needed for the sort interface to compare two Version objects on the
+// slice. If checks if one is less than the other.
+func (c listInfos) Less(i, j int) bool {
+	// Sort tags that aren't semver to front,
+	// this shouldn't happen since we pre filter those to the back
+	if c[i].version == nil {
+		return false
+	} else if c[j].version == nil {
+		return true
+	} else {
+		return c[i].version.LessThan(c[j].version)
+	}
+}
+
+// Swap is needed for the sort interface to replace the Version objects
+// at two different positions in the slice.
+func (c listInfos) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
 var listCommand = &cobra.Command{
-	Use:   "list",
-	Short: "List installed packages",
+	Use:   "list [image]",
+	Short: "List installed packages, or installable images for a package",
 	RunE: func(cmd *cobra.Command, args []string) error {
-		pm := packages.NewPackageManager(viper.GetString("install_path"))
-		packages, err := pm.List()
-		if err != nil {
-			return err
+		ctx := context.Background()
+
+		if arglen := len(args); arglen > 0 {
+			if arglen > 1 {
+				return fmt.Errorf("invalid usage, this command accepts 0 or 1 arguments")
```

Comment:
we should have a better way to command line argument errors, shall we not just run the help in case the arguments does not match expectations?

Reply:
Sure I'll see if i can do something more standard